src/libraries/System.Private.CoreLib/src/System/SearchValues/Strings/Helpers/CharacterFrequencyHelper.cs

```diff
@@ -0,0 +1,127 @@
+        };
+
+        public static void GetSingleStringMultiCharacterOffsets(string value, bool ignoreCase, out int ch2Offset, out int ch3Offset)
+        {
+            Debug.Assert(value.Length > 1);
+            Debug.Assert(!ignoreCase || char.IsAscii(value[0]));
+
+            ch2Offset = IndexOfAsciiCharWithLowestFrequency(value, ignoreCase);
+            ch3Offset = 0;
+
+            if (ch2Offset < 0)
+            {
+                // We have fewer than 2 ASCII chars in the value.
+                Debug.Assert(!ignoreCase);
+
+                // We don't have a frequency table for non-ASCII characters, pick a random one.
+                ch2Offset = value.Length - 1;
+            }
+
+            if (value.Length > 2)
+            {
+                ch3Offset = IndexOfAsciiCharWithLowestFrequency(value, ignoreCase, excludeIndex: ch2Offset);
+
+                if (ch3Offset < 0)
+                {
+                    // We have fewer than 3 ASCII chars in the value.
+                    if (ignoreCase)
+                    {
+                        // We can still use N=2.
+                        ch3Offset = 0;
+                    }
+                    else
+                    {
+                        // We don't have a frequency table for non-ASCII characters, pick a random one.
+                        ch3Offset = value.Length - 1;
+
+                        if (ch2Offset == ch3Offset)
+                        {
+                            ch2Offset--;
+                        }
+                    }
+                }
+            }
+
+            Debug.Assert(ch2Offset != 0);
+            Debug.Assert(ch2Offset != ch3Offset);
+
+            if (ch3Offset > 0 && ch3Offset < ch2Offset)
+            {
+                (ch2Offset, ch3Offset) = (ch3Offset, ch2Offset);
+            }
+        }
+
+        private static int IndexOfAsciiCharWithLowestFrequency(ReadOnlySpan<char> span, bool ignoreCase, int excludeIndex = -1)
+        {
+            float minFrequency = float.MaxValue;
+            int minIndex = -1;
+
+            // Exclude i = 0 as we've already decided to use the first character.
+            for (int i = 1; i < span.Length; i++)
+            {
+                if (i == excludeIndex)
+                {
+                    continue;
+                }
+
+                char c = span[i];
+
+                // We don't have a frequency table for non-ASCII characters, so they are ignored.
+                if (char.IsAscii(c))
+                {
+                    float frequency = AsciiFrequency[c];
+
+                    if (ignoreCase)
+                    {
+                        // Include the alternative character that will also match.
+                        frequency += AsciiFrequency[c ^ 0x20];
+                    }
+
+                    // Avoiding characters from the front of the value for the 2nd and 3rd character
```

Comment:
can you give an example? does this mean that given the 1st char, the 2nd and 3rd char are more predictable than subsequent chars?

Reply:
It would appear that way, yes.
I'm sure there's a lot of fancy logic we could use here to account for things like certain groups of 2 characters being more or less common next to each other than just those characters on their own, etc. I tried to keep it relatively simple.

My approach here was:
- Take all words (length >= 3) from the "The Adventures of Sherlock Holmes" text
- Run this "character & character offset" selection algorithm for each word
- For each word, run it across the whole text and count the number of times we hit a false positive
- False positive is defined as all three characters with these offsets matching, but the input doesn't actually match the whole word at that position

With this text, we're looking at 8360 such words (7779 when ignoring case) and 57306 matches.

A few examples (number of false positives under Ordinal / under OrdinalIgnoreCase):
- With the current PR: 270_568 / 317_342
- Without this `if (i <= 2)` adjustment: 319_310 / 368_302
- Using the first, middle, and last character: 376_549 / 428_664
- Using the first 3 characters: 1_456_934 / 1_630_279
- Using first and last character: 6_474_082 / 7_315_634